docs/Configs.md
# Config essentials

This tutorial explains how to declare a struct to be a config, check for valid configurations, and print configs to string.

**Contents:**
- [Declaring a struct a config](#declaring-a-struct-a-config)
- [Checking for valid configurations](#checking-for-valid-configurations)
- [Printing configs](#printing-configs)

## Declaring a struct a config
Any struct or class can be declared to be a `config_utilities`-config. Simply define a function `void declare_config(ConfigT& config)` and the library will work for `ConfigT`:

```c++
struct MyConfig { ... };
void declare_config(MyConfig& config) { ... }   // Works!
```

> **ℹ️ Note**<br>
> The declaration of `declare_config` *must* be in the same namespace as the object type being declared.

```c++
#include <external/other_object.h>
namespace external {
    void declare_config(OtherObject& config) { ... }  // Also works!
}   // namespace external
```
```c++
#include <external/other_object.h>
void declare_config(external::OtherObject& config) { ... }  // Will not work!
```

In the body of `declare_config`, we tell `config_utilities` everything it needs to know about a config. This consists of a `name` used for human-readable printing and warnings, `fields` and their names to read and write the config, and `checks` that define whether the values stored in config are valid or not:
```c++
void declare_config(MyConfig& config) {
    config::name("MyConfig");
    // Declares the name to be 'MyConfig'. Does not have to match the struct name but is recommended.

    config::field(config.int_val, "int_val_name");
    // Declares that 'MyConfig' has a public member 'int_val' that will be referred to as
    // 'int_val_name' for parsing and printing.

    config::field(config.x, "x", "m");
    // Optionally specify a unit as 3rd argument. This will print that x is in meters.

    config::check(config.x, config::GT, 0, "x");
    // Declares that the config is only valid if config.x > 0. The second "x" defines the name to
    // be printed if the check fails, and is recommended to be the same as the field name.
}
```

## Checking for valid configurations

A variety of checks are supported by `config_utilities`.
Binary checks such as GT (>), GE (>=), LT (<), LE (<=), EQ (==), NE (!=) can be implemented using `check(field, comparison, reference_value, printing_name)`:
```c++
config::check(config.x, config::GT, 0, "x");
```

For a double-sided value check, use `checkInRange(field,, lower, upper, field_name, lower_inclusive, upper_inclusive)`.
E.g., the condition `x is within [0, 100)` becomes:
```c++
  config::checkInRange(config.x, 0.0, 100.0, "x", true, false);
```

Any boolean values can be checked using `config::checkCondition(condition, warning)`:
```c++
config::checkCondition(!config.vector.empty(), "config.vector may not be empty");
```

Custom checks can be implemented and passed as a templated Check Object. This is explained in more details in [ the [advanced features](Advanced.md#adding-custom-checks) tutorial:
```c++
config::check<MyCustomCheck>(custom_arguments);
```

> **ℹ️ Note**<br>
> The above functions are used within `declare_config()` to declare valid values for each field. They *must not* be used outside of it.

To check that a config struct is valid, include `validation.h` and use `isValid(config, print_warnings)`:
```c++
MyConfig cfg;
const bool is_valid = config::isValid(cfg, true);
```
This will print a verbose warning about *all* invalid fields of the config if it fails:
```
=================================== MyConfig ===================================
Warning: Check failed for 'distance': param > 0 (is: '-1').
Warning: Check failed for 'subconfig.f': param within [0, 100) (is: '123').
Warning: Failed to parse param 'vec': Data is not a sequence.
Warning: Failed to parse param 'mat': Incompatible Matrix dimensions: Requested
        3x3 but got 3x4.
Warning: Failed to parse param 'my_enum': Name 'D' is out of bounds for enum wit
        h names ['A', 'B', 'C'].
Warning: Failed to parse param 'uint': Value '-1' underflows storage min of '0'.
================================================================================
```

> **⚠️ Important**<br>
> All failed checks are first formatted by a *formatter* and then printed by a *logger*. If `config_utilities/config_utilities.h` is included these will automatically default to the *asl* formatter and *stdout* logger, resulting in the above print. If no formatter or no logger is included the warning will not show! This holds for all `config_utilities` warnings.

To ensure that a config is valid use `checkValid()`. This will throw and exception or terminate the program if the config is not valid and always print the error.
`checkValid()` return a reference to the config so it can be verified in initializer lists:

```c++
class MyObject {
 public:
  explicit MyObject(const MyConfig& cfg) : config_(config::checkValid(cfg)), object_(config_) {}

private:
  const MyConfig config_;
  ObjectThatNeedsValidConfigs object_;
};
```


## Printing configs

To print configs to string include `printing.h`.
This defines `toString()` for objects declared a config:
```c++
MyConfig cfg;
std::cout << config::toString(cfg) << std::endl;
```

This will print the names and all fields declared in the config:
```
=================================== MyConfig ===================================
i:                            123
distance [m]:                 -0.9876
b:                            false
A ridiculously long field name that will not be wrapped: Short Value (default)
A really really really really really really ridiculously long field name that wi
ll be wrapped:                A really really really ridiculously long string th
                              at will also be wrapped. (default)
vec:                          [5, 4, 3, 2, 1]
map:                          {b: 2, c: 3, d: 4, e: 5}
mat:                          [[1, 2, 3],
                               [4, 5, 6],
                               [7, 8, 9]]
my_enum:                      B
my_strange_enum:              X (default)
sub_config [SubConfig]:
   f:                         55555.55
   s:                         test (default)
   sub_sub_config [SubSubConfig]:
      color:                  [255, 255, 255]
      size:                   101010101
================================================================================
```

> **⚠️ Important**<br>
> Printing calls a *formatter* to format the config contents. If `config_utilities/config_utilities.h` is included this will automatically default to the *asl* formatter, resulting in the above print. If no formatter is included `toString()` will return a warning.

It further defines `operator<<` for all objects declared a config, so this is equivalent:
```c++
std::cout << cfg << std::endl;
```

> **⚠️ Important**<br>
> `operator<<` is only defined in namespace `config` or found by argument-dependent lookup (ADL) in the same namespace as the config object is defined. It is thus generally recommended to use `config::toString()` where possible.


docs/DynamicConfigs.md


# Setup
To use the example GUI, install system deps:

```bash
pip install customtkinter
```


config_utilities/demos/config.cpp
/**
 * TODO(lschmid): Shows how to use configs.
 */

#include "config_utilities/config.h"  // Enables declare_config().

#include <iostream>
#include <string>

#include "config_utilities/formatting/asl.h"         // Simply including this file sets a style to format output.
#include "config_utilities/logging/log_to_stdout.h"  // Simply including this file sets logging to stdout.
#include "config_utilities/parsing/yaml.h"           // Enable fromYamlFile().
#include "config_utilities/printing.h"               // Enable toString()
#include "config_utilities/traits.h"                 // Enables isConfig()
#include "config_utilities/types/eigen_matrix.h"     // Enable parsing and printing of Eigen::Matrix types.
#include "config_utilities/types/enum.h"             // Enable parsing and printing of enum types.
#include "config_utilities/validation.h"             // Enable isValid() and checkValid().

namespace demo {

// A sub-struct for later use.
struct SubSubConfig {
  using Color = Eigen::Matrix<uint8_t, 3, 1>;
  Color color = Color(255, 127, 0);
  size_t size = 5;
};

struct SubConfig {
  float f = 0.123;
  std::string s = "test";
  SubSubConfig sub_sub_config;
};

// A struct that represents what we want to be a config.
// Requirements for a config struct: is default constructable.
struct MyConfig {
  int i = 100;
  std::string s = "test";
  double distance = 42;
  bool b = true;
  std::string test = "Short Value";
  std::string test2 = "A really really really ridiculously long string that will be wrapped.";
  std::string test3 = "A really really really ridiculously long string that will also be wrapped.";
  std::vector<int> vec = {1, 2, 3};
  std::map<std::string, int> map = {{"a", 1}, {"b", 2}, {"c", 3}};
  Eigen::Matrix<double, 3, 3> mat = Eigen::Matrix<double, 3, 3>::Identity();
  enum class MyEnum { kA, kB, kC } my_enum = MyEnum::kA;
  enum MyStrangeEnum : int { kX = 0, kY = 42, kZ = -7 } my_strange_enum = MyStrangeEnum::kX;
  SubConfig sub_config;
};

// Another struct that will not be declared a config.
struct NotAConfig {};

// Defining 'void declare_config(T& config)' function labels a struct as config. All config properties are specified
// within it.
void declare_config(MyConfig& config) {
  // Specify the name.
  config::name("MyConfig");

  // Specify all fields, optionally specifying a unit for readibility..
  config::field(config.i, "i");
  config::field(config.s, "s");
  config::field(config.distance, "distance", "m");
  config::field(config.b, "b");
  config::field(config.test, "A ridiculously long field name that will not be wrapped");
  config::field(config.test2, "A ridiculously long field name that will also not be wrapped");
  config::field(config.test3,
                "A really really really really really really ridiculously long field name that will be wrapped");
  config::field(config.vec, "vec");
  config::field(config.map, "map");
  config::field(config.mat, "mat");

  // String-based enum conversion and verification is supported. Valid values can be specified as a list of names for
  // sequential enums or as a map for non-sequential enums.
  config::enum_field(config.my_enum, "my_enum", {"A", "B", "C"});
  config::enum_field(
      config.my_strange_enum,
      "my_strange_enum",
      {{MyConfig::MyStrangeEnum::kX, "X"}, {MyConfig::MyStrangeEnum::kY, "Y"}, {MyConfig::MyStrangeEnum::kZ, "Z"}});

  // Sub-namespaces can be used to group fields.
  config::enter_namespace("sub_ns");

  // Any other struct that has been declared a config can be a sub-config, declared as a regular field.
  config::field(config.sub_config, "sub_config");

  // Specify all checks to denote a valid configuration. Checks are specified as param, value, and param name to be
  // displayed. Implemented checks are GT (>), GE (>=), LT (<), LE (<=), EQ (==), NE (!=).
  // TODO(lschmid): Would be nice to not duplicate the name but didn't find a nice way to do this.
  config::check(config.i, config::CheckMode::GT, 0, "i");
  config::check(config.i, config::CheckMode::LT, -2, "i");

  // Double sided checks can be invoked as in range.
  const bool lower_inclusive = true;
  const bool upper_inclusive = false;
  config::checkInRange(config.distance, 0.0, 100.0, "distance", lower_inclusive, upper_inclusive);

  // Any other checks can be implmented using the generic condition check.
  config::checkCondition(config.distance < config.i, "param 'distance' must be < 'i'");
  config::checkCondition(!config.s.empty(), "param 's' may not be empty");

  // Any check can be dispatched via templating and forwarded arguments
  config::check<config::internal::CheckRange>(config.distance, 0.0, 100.0, "distance via template");
  config::check<config::internal::Check>(!config.s.empty(), "param 's' may not be empty (via template)");
}

// Declaration of the subconfigs.
void declare_config(SubConfig& config) {
  using namespace config;
  name("SubConfig");
  field(config.f, "f");
  field(config.s, "s");
  field(config.sub_sub_config, "sub_sub_config");
  check(config.f, CheckMode::GT, 0, "f");
}

void declare_config(SubSubConfig& config) {
  using namespace config;
  name("SubSubConfig");
  field(config.color, "color");
  field(config.size, "size");
  check(config.size, CheckMode::EQ, size_t(5), "size");
}

}  // namespace demo

int main(int argc, char** argv) {
  if (argc < 2) {
    std::cerr << "invalid usage! expected resource directory as argument" << std::endl;
    return 1;
  }

  const std::string my_root_path = std::string(argv[1]) + "/";

  // GLobal settings can be set at runtime to change the behavior and presentation of configs.
  config::Settings().inline_subconfig_field_names = true;

  // ===================================== Checking whether a struct is a config =====================================
  std::cout << "\n\n----- Checking whether a struct is a config -----\n\n" << std::endl;

  // Use isConfig<T> to check whether an object has been declared a config.
  std::cout << "MyConfig is a config: " << std::boolalpha << config::isConfig<demo::MyConfig>() << std::endl;
  std::cout << "NotAConfig is a config: " << config::isConfig<demo::NotAConfig>() << std::endl;

  // ====================================== Checking whether a config is valid ======================================
  std::cout << "\n\n----- Checking whether a config is valid \n\n" << std::endl;

  // Create a valid and an invalid config.
  demo::MyConfig config, invalid_config;
  invalid_config.i = -1;
  invalid_config.distance = 123;
  invalid_config.s.clear();
  invalid_config.sub_config.f = -1.f;
  invalid_config.sub_config.sub_sub_config.size = 0;

  // Print whether they are valid. Since we invalidated all fields of 'invalid_config' a comprehensive summary of all
  // issues is printed.
  constexpr bool print_warnings = true;
  const bool config_is_valid = config::isValid(config, print_warnings);
  std::cout << "'config' is valid: " << config_is_valid << std::endl;

  const bool invalid_config_is_valid = config::isValid(invalid_config, print_warnings);
  std::cout << "'invalid_config' is valid: " << invalid_config_is_valid << std::endl;

  // Check valid will enforce that the config is valid, throwing an error and always printing the warnings if not.
  try {
    config::checkValid(invalid_config);
  } catch (std::runtime_error& e) {
    std::cout << "Exception thrown: " << e.what() << std::endl;
  }

  // ======================================== Read the config from file ========================================
  std::cout << "\n\n----- Reading the config from file -----\n\n" << std::endl;

  // Read the config from file.
  config = config::fromYamlFile<demo::MyConfig>(my_root_path + "demo_params.yaml");

  std::cout << "Read values i='" << config.i << "', s='" << config.s << "', distance='" << config.distance
            << "' from file." << std::endl;
  std::cout << "Enum 'config.my_enum' is now B: " << (config.my_enum == demo::MyConfig::MyEnum::kB) << std::endl;

  // Any errors parsing configs will print verbose warnings if desired and use the default values.
  invalid_config = config::fromYamlFile<demo::MyConfig>(my_root_path + "demo_invalid_params.yaml");

  // ======================================== Printing configs to string ========================================
  std::cout << "\n\n----- Printing configs to string -----\n\n" << std::endl;

  // Easier automatic printing of all configs with unit and additional information can be done using the toString():
  const std::string config_as_string = config::toString(config);
  std::cout << config_as_string << std::endl;

  // Inclunding "printing.h" also implements th ostream operator for decared config types. The above is thus equivalent
  // to:
  std::cout << config << std::endl;

  return 0;
}
